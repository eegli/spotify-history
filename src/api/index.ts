import axios, { AxiosRequestConfig } from 'axios';

export enum Endpoints {
  RefreshToken = 'https://accounts.spotify.com/api/token',
  GetHistory = 'https://api.spotify.com/v1/me/player/recently-played',
}

export interface GetHistoryParams {
  limit?: number;
  after?: number;
  before?: number;
}

export interface GetHistoryResponse
  extends SpotifyApi.UsersRecentlyPlayedTracksResponse {}

export interface GetRefreshTokenParams {
  client_id: string;
  client_secret: string;
  grant_type: 'refresh_token';
  refresh_token: string;
}

export interface GetRefreshTokenResponse {
  access_token: string;
  token_type: string;
  scope: string;
  expires_in: number;
}

export function instance<T>(params?: T) {
  return axios.create({
    params,
  });
}
