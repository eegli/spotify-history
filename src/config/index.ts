const { env } = process;

type Config = {
  TABLE_NAME: string;
  masterDateRef: string;
  SPT_CLIENT_ID: string;
  SPT_CLIENT_SECRET: string;
  SPT_REFRESH_TOKEN: string;
  AWS_REGION: string;
};

const config: Readonly<Config> = {
  TABLE_NAME: env.TABLE_NAME || '',
  masterDateRef: '1970-01-01T00:00:00.001Z',
  AWS_REGION: env.AWS_REGION || '',
  SPT_CLIENT_ID: env.CLIENT_ID || '',
  SPT_CLIENT_SECRET: env.CLIENT_SECRET || '',
  SPT_REFRESH_TOKEN: env.REFRESH_TOKEN || '',
};

export default config;
