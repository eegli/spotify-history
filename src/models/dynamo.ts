interface DynamoBaseDoc {
  TableName: string;
}

// Primary key for this table
// Optional sort key may be added here
interface DynamoBaseKeys {
  dateId: string;
}

// For GET operations
export interface DynamoRef extends DynamoBaseDoc {
  Key: DynamoBaseKeys;
}

// For PUT operations
export interface DynamoItem extends DynamoBaseDoc {
  Item: DynamoBaseKeys & DynamoRecordAttrs;
}

// All non-required item attributes
interface DynamoRecordAttrs {
  lastScrobbed: string;
  lastScrobbedString: string;
  songs?: any;
}
