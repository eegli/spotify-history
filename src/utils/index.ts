import axios, { AxiosError } from 'axios';
import { writeFileSync } from 'fs';

export const isAxiosError = (err: unknown): err is AxiosError =>
  axios.isAxiosError(err);

export type PickType<T, K extends keyof T> = T[K];

export const write = (data: unknown, fileName: string): void => {
  writeFileSync(fileName + '.json', JSON.stringify(data));
};

export const backupFileNameDates = (date: Date) => {
  // Leading zero for month
  const m = date.getMonth();
  const month = m < 10 ? '0' + m : m.toString();

  // Get number of week
  const firstDayOfYear = new Date(date.getFullYear(), 0, 1);
  const pastDaysOfYear = (date.getTime() - firstDayOfYear.getTime()) / 86400000;
  const week = Math.ceil((pastDaysOfYear + firstDayOfYear.getDay() + 1) / 7);

  // Year, month, week
  return [date.getFullYear(), month, week];
};

// https://stackoverflow.com/a/18650828
export const fileSizeFormat = (bytes: number, decimals = 2) => {
  if (bytes === 0) return '0 Bytes';

  const k = 1024;
  const dm = decimals < 0 ? 0 : decimals;
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
};
