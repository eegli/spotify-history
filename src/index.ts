import Hello from './module/hello';

import { ScheduledHandler } from 'aws-lambda';

import { config } from './config';
const { env } = process;

import mongoose, { ConnectOptions } from 'mongoose';

const pw = env.MONGO_PW || '';
const user = env.MONGO_USER || '';

const m = new mongoose.Mongoose();
const options: ConnectOptions = {
  useNewUrlParser: true,
  useUnifiedTopology: true,
  dbName: config.dbName,
  serverSelectionTimeoutMS: 5000,
};

async function run(): Promise<void> {
  try {
    await m.connect(
      `mongodb+srv://${user}:${pw}@spotify-history-stg.rfb2f.mongodb.net/?retryWrites=true&w=majority`,
      options
    );
    const db = m.connection.collection('movies');
    // Query for a movie that has the title 'Back to the Future'
    const query = { title: 'Back to the Future' };
    const movie = await db.findOne(query);
    console.log(movie);
  } catch (err) {
    throw new Error(err);
  } finally {
    // Ensures that the client will close when you finish/error
    await m.disconnect();
  }
}

export const handler: ScheduledHandler = async (
  event,
  context,
  callback
): Promise<void> => {
  run()
    .then(() => callback(null))
    .catch(err => callback(new Error(err)));

  /*   console.info(Hello + `, current env is: ${process.env.STAGE}`);
  console.warn(`Available in both prod and stg: ${process.env.PUBLIC_TEST}`);
  console.log('EVENT' + JSON.stringify(event, null, 2));
  console.log('CONTEXT' + JSON.stringify(context, null, 2)); */
};
