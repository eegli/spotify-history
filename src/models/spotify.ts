import axios from 'axios';
import { URLSearchParams } from 'url';
import config from '../config';

interface RefreshTokenRes {
  access_token: string;
  token_type: string;
  scope: string;
  expires_in: number;
}

export interface HistoryParams {
  limit?: number;
  after?: number;
  before?: number;
}

type HistoryResponse = SpotifyApi.UsersRecentlyPlayedTracksResponse;

type HistoryMeta = Pick<
  SpotifyApi.CursorBasedPagingObject<SpotifyApi.PlayHistoryObject>,
  'cursors'
>;

type History = SpotifyApi.PlayHistoryObject[];

interface TrackObjEnriched {
  name: string;
  id: string;
  artists: {
    artistName: string;
    artistId: string;
    genres: string;
  }[];
}

export class Spotify {
  bearerToken: string = '';
  history: History = [];
  historyMeta: HistoryMeta = {
    cursors: {
      after: '',
      before: '',
    },
  };

  async getRefreshToken(): Promise<void> {
    const params = new URLSearchParams();
    params.append('client_id', config.sptClientId);
    params.append('client_secret', config.sptClientSecret);
    params.append('grant_type', 'refresh_token');
    params.append('refresh_token', config.sptRefreshToken);

    const res = await axios.post<RefreshTokenRes>(
      'https://accounts.spotify.com/api/token',
      params,
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      }
    );
    console.log(res);
    this.bearerToken = res.data.access_token || '';
  }

  async getHistory(params: HistoryParams): Promise<void> {
    const res = await axios.get<HistoryResponse>(
      'https://api.spotify.com/v1/me/player/recently-played',
      {
        headers: {
          Authorization: `Bearer ${this.bearerToken}`,
        },
        params,
      }
    );
    this.history = res.data.items;
  }

  // https://dev.to/pedrohasantiago/typescript-adjusting-types-in-reduce-function-with-an-async-callback-2kc8
  async enrichHistory() {
    return await this.history.reduce(async (acc, el) => {
      const genre = await axios.get<SpotifyApi.MultipleArtistsResponse>(
        `https://api.spotify.com/v1/artists`,
        {
          params: {
            ids: el.track.artists.map(artist => artist.id).join(','),
          },
          headers: {
            Authorization: `Bearer ${this.bearerToken}`,
          },
        }
      );
      console.log(genre);
      acc
        .then(a =>
          a.push({
            name: el.track.name,
            id: el.track.id,
            artists: genre.data.artists.map(el => ({
              artistName: el.name,
              artistId: el.id,
              genres: el.genres.join(','),
            })),
          })
        )
        .catch(err => console.log(JSON.stringify(err)));
      return acc;
    }, Promise.resolve(<TrackObjEnriched[]>[]));
  }
}
