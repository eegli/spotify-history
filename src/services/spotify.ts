import { initialize, SpotifyClient } from '@spotifly/core';
import config from '../config';
import { DynamoHistoryElement } from '../config/types';

export type HistoryParams = Parameters<
  SpotifyClient['Player']['getRecentlyPlayedTracks']
>[0];

export interface RefreshTokenResponse {
  access_token: string;
  token_type: string;
  scope: string;
  expires_in: number;
}

// Export so we can use it in the tests
export type HistoryResponse = SpotifyApi.UsersRecentlyPlayedTracksResponse;
export type MultipleArtistsResponse = SpotifyApi.MultipleArtistsResponse;

export default class Spotify {
  private client: SpotifyClient;
  private items: SpotifyApi.PlayHistoryObject[] = [];
  // Unix timestamps
  cursorBefore = 0;
  cursorAfter = 0;

  constructor() {
    this.client = initialize({
      clientId: config.SPOTIFY.client_id,
      clientSecret: config.SPOTIFY.client_secret,
      refreshToken: config.SPOTIFY.refresh_token,
    });
  }

  get count() {
    return this.items.length;
  }

  // Gets the raw history from Spotify
  async fetchHistory(params: HistoryParams): Promise<void> {
    const res = await this.client.Player.getRecentlyPlayedTracks({
      ...params,
      limit: 50,
    });

    if (res.data.items.length > 0) {
      this.items = res.data.items;

      const firstElem = this.items[0].played_at;
      const lastElem = this.items[this.items.length - 1].played_at;

      // Manually set the cursors so that we know the time of the
      // first and last song of this history
      this.cursorAfter = new Date(lastElem).getTime();
      this.cursorBefore = new Date(firstElem).getTime();
    }
  }

  // Create the actual history that will be saved in dynamo
  createHistory(): DynamoHistoryElement[] {
    return this.items.reduce((acc, el) => {
      acc.push({
        name: el.track.name,
        id: el.track.id,
        playedAt: new Date(el.played_at).toISOString(),
      });

      return acc;
    }, <DynamoHistoryElement[]>[]);
  }
}
