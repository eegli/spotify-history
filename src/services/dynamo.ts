import config from '../config';

import { DataMapper, QueryOptions } from '@aws/dynamodb-data-mapper';
import { ConditionExpression } from '@aws/dynamodb-expressions';
import DynamoDB from 'aws-sdk/clients/dynamodb';

import { ServiceConfigurationOptions } from 'aws-sdk/lib/service';
import { History, HistoryElement } from '../models/history';

type Options = DynamoDB.DocumentClient.DocumentClientOptions &
  ServiceConfigurationOptions &
  DynamoDB.ClientApiVersions;

const params: Options = {
  region: config.AWS_REGION,
  apiVersion: '2012-08-10',
};

if (process.env.stage === 'local') {
  params.region = 'localhost';
  params.endpoint = 'http://localhost:8000';
}

const mapper = new DataMapper({
  client: new DynamoDB(params),
});

export const dynamoGetLatestHistory = async () => {
  const queryOptions: QueryOptions = {
    limit: 1,
    scanIndexForward: false,
  };

  const params: ConditionExpression = {
    type: 'Equals',
    subject: 'type',
    object: 'history',
  };
  const iterator = mapper.query(History, params, queryOptions);
  for await (const history of iterator) {
    // Return the first element
    return history;
  }
};

export const dynamoSetHistory = async (
  timestamp: string,
  count: number,
  songs: HistoryElement[]
) => {
  const newHistory: History = Object.assign(new History(), {
    timestamp,
    count,
    songs,
  });

  return mapper.put(newHistory);
};
