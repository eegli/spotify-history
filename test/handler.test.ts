import { Context, EventBridgeEvent } from 'aws-lambda';
import { QueryIterator, StringToAnyObjectMap } from '@aws/dynamodb-data-mapper';
import { handler } from '../src/';
import {
  rawSpotifyHistory,
  enrichedSpotifyHistory,
  dynamoData,
} from './payloads';
import { mocked } from 'ts-jest/utils';

const mockFetchSpotifyData = jest.fn();
const mockCreateHistory = jest.fn().mockResolvedValue(enrichedSpotifyHistory());

jest.mock('@aws/dynamodb-data-mapper');
jest.mock('../src/services/spotify', () => {
  return {
    Spotify: jest.fn().mockImplementation(() => {
      return {
        getRefreshToken: jest.fn(),
        fetchSpotifyData: mockFetchSpotifyData,
        createHistory: mockCreateHistory,
        items: rawSpotifyHistory(),
      };
    }),
  };
});

import { mapper } from '../src/services/dynamo';
import { Spotify } from '../src/services/spotify';

const mockedSpotify = mocked(Spotify, true);
const querySpy = jest.spyOn(mapper, 'query');
const putSpy = jest.spyOn(mapper, 'put');

const iteratorMock = jest.fn();
// @ts-ignore
const qi: QueryIterator<StringToAnyObjectMap> = QueryIterator;
// @ts-ignore
qi[Symbol.iterator] = iteratorMock;

querySpy.mockReturnValue(qi);

beforeEach(() => {
  querySpy.mockClear();
  putSpy.mockClear();
  iteratorMock.mockClear();
});

describe('Handler', () => {
  it('Should run without errors', async () => {
    iteratorMock.mockReturnValueOnce([].values());
    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
    expect(mockedSpotify.mock.instances.length).toEqual(1);
    expect(querySpy).toHaveBeenCalledTimes(1);
    expect(putSpy).toHaveBeenCalledTimes(1);
    expect(putSpy.mock.calls).toMatchSnapshot();
  });

  it('Should query for all songs if there are no existing ones', async () => {
    iteratorMock.mockReturnValueOnce(dynamoData().values());

    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
  });

  it('Should query for new songs only if there are existing ones', async () => {
    iteratorMock.mockReturnValueOnce([].values());

    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
  });
});
