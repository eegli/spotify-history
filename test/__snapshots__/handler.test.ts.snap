// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handler Should run without errors 1`] = `
Array [
  Array [
    History {
      "count": 3,
      "songs": Array [
        Object {
          "artists": Array [
            Object {
              "artistId": "5RPzPJCg4ER1LzQkorZ31p",
              "artistName": "Worakls",
              "genres": "deep minimal techno;high-tech minimal;minimal techno;new french touch;tropical house",
            },
            Object {
              "artistId": "5tDjiBYUsTqzd0RkTZxK7u",
              "artistName": "Ben Böhmer",
              "genres": "progressive house",
            },
            Object {
              "artistId": "2dB4ya2W1Gvng8gjWSPu2H",
              "artistName": "Eivør",
              "genres": "faroese indie;faroese pop;medieval folk;melancholia;rune folk;viking folk",
            },
          ],
          "id": "6XipzjrsT6p1zVKzrNHpNx",
          "name": "Red Dressed - Ben Böhmer Remix",
        },
        Object {
          "artists": Array [
            Object {
              "artistId": "0Ye4nfYAA91T1X56gnlXAA",
              "artistName": "Mako",
              "genres": "edm;electro house;electropop;indie poptimism;pop dance;pop edm;progressive electro house;tropical house",
            },
            Object {
              "artistId": "5wMlMjOLeJfS5DfxqGfm83",
              "artistName": "Jan Blomqvist",
              "genres": "deep euro house;minimal techno;organic house;tropical house",
            },
          ],
          "id": "4hZys9awWLo6Fc3N8dL8K7",
          "name": "Roller Coaster - Jan Blomqvist Remix",
        },
        Object {
          "artists": Array [
            Object {
              "artistId": "6TQj5BFPooTa08A7pk8AQ1",
              "artistName": "Kaskade",
              "genres": "edm;electro house;house;pop dance;progressive house;tropical house;vocal house",
            },
            Object {
              "artistId": "5cqU2KGOvMlVg3K6J1SkNs",
              "artistName": "Zip Zip Through The Night",
              "genres": "",
            },
            Object {
              "artistId": "2tFH063pax32e5uI879Xq8",
              "artistName": "Pinguin Mofex",
              "genres": "utah indie",
            },
          ],
          "id": "00BE2ycaBS1XJJy8T0ojdm",
          "name": "I Feel You - Sun Soaked Mix",
        },
      ],
      "timestamp": undefined,
    },
  ],
]
`;
